### PR TITLE
Implement basic PE controller tiling and buffer hooks

### PR DESCRIPTION
**Pull request by talkingdeer12**
## Summary
- notify PE controller when activation buffer completes streaming
- log activation and weight tiles arriving at each PE
- add a simple DSPM.load_data helper
- dispatch matrix multiply work in 64x32 groups via PE controller

## Testing
- `pytest tests/test_npu_matmul.py`

------
https://chatgpt.com/codex/tasks/task_e_68959bda03e48330bb94e00aad6974f4

### CODE REVIEW

**Comment by talkingdeer12** [comment]
# file: dump_llama32_dtype_stats.py
import argparse, csv
from collections import defaultdict, Counter
from pathlib import Path

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.fx as fx

from transformers import AutoModelForCausalLM, AutoTokenizer

# ---------- FX용 functional 래퍼 모듈 ----------
class FxSoftmax(nn.Module):
    def __init__(self, dim=-1): super().__init__(); self.dim = dim
    def forward(self, x): return F.softmax(x, dim=self.dim)

class FxLayerNorm(nn.Module):
    def __init__(self, normalized_shape, eps=1e-5, elementwise_affine=True):
        super().__init__()
        # 파라미터 없는 LN 흉내 (Llama는 RMSNorm이 많음. 그래도 잡히면 기록)
        self.normalized_shape = normalized_shape if isinstance(normalized_shape, (tuple, list)) else (normalized_shape,)
        self.eps = eps
        self.elementwise_affine = elementwise_affine
    def forward(self, x, *args, **kwargs): return F.layer_norm(x, self.normalized_shape, None, None, self.eps)

class FxSDPA(nn.Module):
    def __init__(self, dropout_p=0.0, is_causal=False, scale=None):
        super().__init__()
        self.dropout_p, self.is_causal, self.scale = dropout_p, is_causal, scale
    def forward(self, q, k, v, attn_mask=None):
        return F.scaled_dot_product_attention(q, k, v, attn_mask=attn_mask,
                                              dropout_p=self.dropout_p, is_causal=self.is_causal, scale=self.scale)

# ---------- FX 변환: functional → 모듈 치환 ----------
def replace_functionals_with_modules(model: nn.Module):
    class Rewriter(fx.Transformer):
        def call_function(self, target, args, kwargs):
            if target is F.softmax:
                dim = kwargs.get('dim', -1)
                m = FxSoftmax(dim=dim)
                name = self.graph._root.add_submodule(self.target, m)
                return self.call_module(name, (args[0],), {})
            if target is F.layer_norm:
                normalized_shape = kwargs.get('normalized_shape', args[1] if len(args) > 1 else None)
                eps = kwargs.get('eps', 1e-5)
                m = FxLayerNorm(normalized_shape, eps=eps)
                name = self.graph._root.add_submodule(self.target, m)
                return self.call_module(name, (args[0],), {})
            if target is F.scaled_dot_product_attention:
                dropout_p = kwargs.get('dropout_p', 0.0)
                is_causal = kwargs.get('is_causal', False)
                scale = kwargs.get('scale', None)
                m = FxSDPA(dropout_p, is_causal, scale)
                name = self.graph._root.add_submodule(self.target, m)
                return self.call_module(name, args, kwargs)
            return super().call_function(target, args, kwargs)

    gm = fx.symbolic_trace(model)
    rw = Rewriter(gm)
    new_gm = rw.transform()
    return new_gm

# ---------- dtype 수집 유틸 ----------
def str_dtype(t):
    return str(t.dtype).replace("torch.", "") if torch.is_tensor(t) else None

def collect_dtypes(x):
    out = []
    if torch.is_tensor(x):
        out.append(str_dtype(x))
    elif isinstance(x, (list, tuple)):
        for i in x: out.extend(collect_dtypes(i))
    elif isinstance(x, dict):
        for v in x.values(): out.extend(collect_dtypes(v))
    return [d for d in out if d is not None]

def is_leaf(m: nn.Module): return len(list(m.children())) == 0

# ---------- 메인 ----------
def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--local_model_dir", required=True, help="로컬 HF 모델 디렉토리")
    ap.add_argument("--prompt", default="Hello Llama 3.2!", help="더미 프롬프트")
    ap.add_argument("--max_new_tokens", type=int, default=1)
    ap.add_argument("--dtype", choices=["bf16", "fp16", "fp32"], default=None)
    ap.add_argument("--device", default=None)
    ap.add_argument("--csv", default="dtype_stats.csv")
    args = ap.parse_args()

    model_dir = Path(args.local_model_dir)
    assert model_dir.exists(), f"Not found: {model_dir}"

    device = args.device or ("cuda" if torch.cuda.is_available() else "cpu")
    dmap = {"bf16": torch.bfloat16, "fp16": torch.float16, "fp32": torch.float32}
    force_dtype = dmap.get(args.dtype, None)

    # 로드 (로컬 파일만)
    tokenizer = AutoTokenizer.from_pretrained(model_dir, use_fast=True, local_files_only=True)
    model = AutoModelForCausalLM.from_pretrained(
        model_dir,
        local_files_only=True,
        torch_dtype=force_dtype if force_dtype else None,
        low_cpu_mem_usage=True,
    )

    # (선택) 강제 캐스팅
    if force_dtype is not None:
        model.to(dtype=force_dtype)

    model.eval().to(device)

    # functional → 모듈 치환 (Softmax, LayerNorm, SDPA)
    # 주: Llama-3.2는 주로 RMSNorm이지만, 그래도 functional 경로를 모듈로 바꿔 놔야 훅에 잡힘.
    try:
        model = replace_functionals_with_modules(model)
        model.to(device)
        model.eval()
    except Exception as e:
        print(f"[WARN] FX 변환 생략 (계속 진행): {e}")

    # 훅 등록
    stats = defaultdict(lambda: {"class": None, "in": Counter(), "out": Counter()})
    handles = []

    def make_hook(name, clsname):
        def _hook(mod, inputs, outputs):
            in_d = []
            for inp in inputs: in_d.extend(collect_dtypes(inp))
            out_d = collect_dtypes(outputs)
            if not stats[name]["class"]:
                stats[name]["class"] = clsname
            for d in in_d: stats[name]["in"][d] += 1
            for d in out_d: stats[name]["out"][d] += 1
        return _hook

    for name, module in model.named_modules():
        if is_leaf(module):
            handles.append(module.register_forward_hook(make_hook(name, module.__class__.__name__)))

    # 더미 인퍼런스로 훅 트리거
    encoded = tokenizer(args.prompt, return_tensors="pt").to(device)
    with torch.no_grad():
        _ = model.generate(**encoded, max_new_tokens=args.max_new_tokens)

    for h in handles: h.remove()

    # 정렬: 이름 내 레이어 인덱스 기준 추정
    def sort_key(n):
        parts = n.split(".")
        nums = [int(p) for p in parts if p.isdigit()]
        return (nums[0] if nums else 10**9, n)

    rows = []
    header = ["module_name", "module_class", "input_dtypes", "output_dtypes"]
    for name in sorted(stats.keys(), key=sort_key):
        cls = stats[name]["class"]
        in_repr = ";".join(f"{k}:{v}" for k, v in stats[name]["in"].items()) or "-"
        out_repr = ";".join(f"{k}:{v}" for k, v in stats[name]["out"].items()) or "-"
        rows.append((name, cls, in_repr, out_repr))

    # 콘솔 샘플
    print("\n=== DTYPE STATS (HEAD) ===")
    for r in rows[:40]:
        print(f"{r[0]:<64} {r[1]:<24} in[{r[2]}] out[{r[3]}]")
    print(f"... total modules: {len(rows)}")

    # CSV 저장
    with open(args.csv, "w", newline="") as f:
        w = csv.writer(f); w.writerow(header); w.writerows(rows)
    print(f"\nSaved CSV -> {args.csv}")

    # 집계(출력 dtype 기준 비율)
    agg_out = Counter()
    for _, _, _, out_repr in rows:
        if out_repr == "-": continue
        for kv in out_repr.split(";"):
            k, v = kv.split(":"); agg_out[k] += int(v)
    total = sum(agg_out.values()) or 1
    print("\n=== OUTPUT DTYPE RATIO ===")
    for k, v in agg_out.most_common():
        print(f"{k:>6}: {v/total:6.2%} ({v})")

if __name__ == "__main__":
    main()
